packages/expo-permissions/android/src/test/java/expo/modules/permissions/PermissionsServiceTest.kt

```diff
@@ -0,0 +1,174 @@
+import io.mockk.verify
+import io.mockk.verifyAll
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+import org.unimodules.interfaces.permissions.PermissionsResponse
+import org.unimodules.interfaces.permissions.PermissionsResponseListener
+import org.unimodules.interfaces.permissions.PermissionsStatus
+
+private val ANDROID_PERMISSIONS = arrayOf(
+    Manifest.permission.CAMERA,
+    Manifest.permission.READ_CALENDAR,
+    Manifest.permission.WRITE_CALENDAR,
+    Manifest.permission.RECORD_AUDIO,
+    Manifest.permission.READ_SMS,
+    Manifest.permission.ACCESS_COARSE_LOCATION,
+    Manifest.permission.ACCESS_FINE_LOCATION
+)
+
+@RunWith(JUnit4::class)
+class PermissionsServiceTest {
+
+  private val mActivityDelegate = mockk<PermissionsActivityDelegate>()
+  private val mPermissionsCache = mockk<PermissionCache>()
+  private val mContext = mockk<Context>()
+
+  private fun generatePermissionsResponse(permissions: Array<String>, status: PermissionsStatus, canAskAgain: Boolean = true): Map<String, PermissionsResponse> {
+    return permissions.zip(Array(permissions.size) { PermissionsResponse(status, canAskAgain) }).toMap()
+  }
+
+  @Before
+  fun resetMock() {
+    clearMocks(mActivityDelegate, mPermissionsCache, mContext)
+  }
+
+  @Test
+  fun `gets permissions for the first time`() {
+    getAndVerifyPermissions(ANDROID_PERMISSIONS, PermissionsStatus.UNDETERMINED, systemStatus = false, wasAsked = false, canAskAgain = true)
+  }
+
+  @Test
+  fun `gets granted permissions`() {
+    getAndVerifyPermissions(ANDROID_PERMISSIONS, PermissionsStatus.GRANTED, systemStatus = true, wasAsked = false, canAskAgain = true)
+  }
+
+  @Test
+  fun `gets denied permissions`() {
+    getAndVerifyPermissions(ANDROID_PERMISSIONS, PermissionsStatus.DENIED, systemStatus = false, wasAsked = true, canAskAgain = true)
+  }
+
+  @Test
+  fun `gets denied permissions with canAksAgain set to false`() {
+    getAndVerifyPermissions(ANDROID_PERMISSIONS, PermissionsStatus.DENIED, systemStatus = false, wasAsked = true, canAskAgain = false)
+  }
+
+  @Test
+  fun `requests permissions and grants them`() {
+    requestAndVerifyPermissions(ANDROID_PERMISSIONS, PermissionsStatus.GRANTED, userResponse = true, canAskAgain = true)
+  }
+
+
+  @Test
+  fun `requests permissions and denies them but can ask again`() {
+    requestAndVerifyPermissions(ANDROID_PERMISSIONS, PermissionsStatus.DENIED, userResponse = false, canAskAgain = true)
+  }
+
+  @Test
+  fun `requests permissions and denies them but can't ask again`() {
+    requestAndVerifyPermissions(ANDROID_PERMISSIONS, PermissionsStatus.DENIED, userResponse = false, canAskAgain = false)
+  }
+
+  private fun requestAndVerifyPermissions(permissions: Array<String>, expectedStatus: PermissionsStatus,
+                                          userResponse: Boolean, canAskAgain: Boolean) {
+    every { mActivityDelegate.askForPermissions(any(), any()) } answers {
```

Comment:
I generally like the concept and approve it, but maybe you would consider one more improvement?

Tests are often written in `given->when-then` convention. Each part is seperated from the other, given might be ommited.
Separation comes either by formatting and comment between each part, or putting each part into seperate function. Hence, example test might look like that:
```kt
fun test(){
  givenPermission(name = "permissionName", isDenied = true, canAskAgain = false)
  whenAskingForPermission()

  //then
  verify { thatSomethingHappened }
}
```

This is just suggestion :)